dvc/repo/run.py

```diff
@@ -1,23 +1,39 @@
 from __future__ import unicode_literals
 
-from dvc.repo.scm_context import scm_context
+import os
+from shortuuid import uuid
 
+from .ctx import ctx
 
-@scm_context
+
+@ctx
 def run(self, no_exec=False, **kwargs):
     from dvc.stage import Stage
 
-    with self.state:
-        stage = Stage.create(self, **kwargs)
+    stage = Stage.create(self, **kwargs)
 
     if stage is None:
         return None
 
     self.check_dag(self.stages() + [stage])
 
-    with self.state:
+    # We are releasing locks in stage.run() to allow parallel execution. To
+    # avoid possible DAG problems, we are creating the dvc file now, so
+    # that other `dvc run`s could use it to verify DAG on their side before
+    # launching their commands. We are using random unique name for this
+    # temporary dvc file to avoid collisions in case someone is trying to
+    # create the same dvc file in another process.
+    dname = os.path.dirname(stage.path)
+    tmp_fname = os.path.join(
+        dname, "." + str(uuid()) + stage.STAGE_FILE_SUFFIX
+    )
+    stage.dump(fname=tmp_fname)
```

Comment:
Not sure using random name is a good idea. This will lead to a silent overwrite later. Why don't use it original name from the start? it will prompt user, which looks more sane behavior.